cohortextractor/dsl.py

```diff
@@ -0,0 +1,178 @@
+
+
+class Cohort:
+    """Represents the cohort of patients in a study."""
+
+    def set_population(self, population: PatientSeries) -> None:
+        """Set the population variable for this cohort."""
+
+    def add_variable(self, name: str, variable: PatientSeries) -> None:
+        """Add a variable to this cohort by name."""
+
+    def __setattr__(self, name: str, variable: PatientSeries) -> None:
+        return self.add_variable(name, variable)
+
+
+class EventFrame:
+    """Represents a collection of records, with multiple rows per patient.
+
+    Either an EventTable, or the result of filtering an EventTable.
+    """
+
+    def filter(self, filter: Expression) -> EventFrame:  # noqa: A002, A003
+        """Return a new EventFrame with given filter."""
+
+    def sort_by(self, *columns: ColumnOrName) -> SortedEventFrame:
+        """Return a SortedEventFrame with given sort column."""
+
+    def count(self) -> PatientSeries:
+        """Return a PatientSeries with count of matching events per patient."""
+
+    def exists(self) -> PatientSeries:
+        """Return a PatientSeries indicating whether each patient has a matching event."""
+
+    def __getattr__(self, name: str) -> NoReturn:
+        ...
+
+
+class SortedEventFrame:
+    """Represents an EventFrame that has been sorted."""
+
+    def first(self) -> PatientFrame:
+        """Return a PatientFrame with the first event for each patient."""
+
+    def last(self) -> PatientFrame:
+        """Return a PatientFrame with the last event for each patient."""
+
+    def __getattr__(self, name: str) -> NoReturn:
+        ...
+
+
+class PatientFrame:
+    """Represents a collection of records, with one row per patient.
+
+    Either a PatientTable, or the result of filtering a PatientTable.
+    """
+
+    def filter(self, filter: Expression) -> PatientFrame:  # noqa: A002, A003
+        """Return a new PatientFrame with given filter.
+
+        >>> filtered = table.filter(table.code in codes)
+        """
+
+    def select_column(self, column: ColumnOrName) -> PatientSeries:
+        """Return a PatientSeries containing given column.
+
+        >>> column = table.select_column(table.date)
+        >>> column = table.select_column("date")
+        """
+
+    def __getattr__(self, name: str) -> NoReturn:
+        ...
+
+
+class PatientSeries:
+    """Represents a column indexed by patient.
+
+    Can be used as a variable in a Cohort, or as an input when computing another
+    variable.
+    """
+
+    def __eq__(self, other: Expression) -> PatientSeries:  # type: ignore
+        """Return Expression indicating whether self is equal to other."""
+
+    def __ne__(self, other: Expression) -> PatientSeries:  # type: ignore
+        ...
+
+    def __add__(self, other: Expression) -> PatientSeries:
+        ...
+
+    def __gt__(self, other: Expression) -> PatientSeries:
+        ...
+
+    # etc
+
+    def __getattr__(self, name: str) -> NoReturn:
+        ...
```

Comment:
Can we type-parameterize this like

```py
class PatientSeries(Generic[T]):
    def __eq__(self, other: Expression) -> PatientSeries[bool]:
        ...

    def __add__(self, other: Expression) -> PatientSeries[T]:
        ...
```

Reply:
Yes, I wondered about that.  I'm not sure how we do things like ensure that you can only do `s1 + s2` when both `s1` and `s2` are `PatientSeries[Num => T]`.  Possibly with `overload` or `TypeGuard`?  Would need to investigate.